Compatibility clearout

johnhaddon opened the pull request:
Our general approach to providing backwards compatibility for old files and scripts is to "monkey patch" the Python API using config files that are loaded at startup. This has been relatively successful, with the major benefit being much cleaner core code, because the compatibility hacks are largely isolated in configs.

Over time though, these configs have built up considerable cruft and in some cases even have an adverse effect on script load times. This PR takes what I think is a reasonably conservative approach to retiring the older configs, with the intention being that it signals the start of a continuous process of cruft removal in the future. As noted in the Changes file, the absolute worst case scenario is that places with a long history of Gaffer usage will need to adopt some of these configs internally :

- Removed several compatibility shims for loading files from Gaffer versions prior to 0.20.0.0. Either resave the file from version 0.57.0.0 or adopt the appropriate shim into your own configuration. The following are affected :
     - RemoveChannels nodes.
     - OpenColorIO nodes.
     - Certain ImageWriter settings from version 0.17.0.0 and prior.
     - Expression nodes from version 0.18.0.0 and prior.
     - Invalid `Node.user` plug inputs from version 0.18.0.0 and prior.
     - Reformat nodes from version 0.19.0.0 and prior.
     - Crop nodes from version 0.20.0.0 and prior.
     - TaskNodes from version 0.20.0.0 and prior.
- Removed several compatibility shims for running old Python code without modification. Either update your code or adopt the appropriate shim into your own configuration.
    - `Gaffer.MetadataAlgo` and `Gaffer.MonitorAlgo` can no longer be accessed without namespace qualification.
    - `GafferImage.ImageAlgo` and `GafferImage.BufferAlgo` can no longer be accessed without namespace qualification.
    - `GafferScene.SceneAlgo` can no longer be accessed without namespace qualification.
    - `GafferDispatch` attributes can no longer be accessed via the `Gaffer` namespace.
    - `Gaffer.UndoContext` is no longer an alias for `Gaffer.UndoScope`.
    - `GafferDispatchUI` attributes can no longer be accessed via the `GafferUI` namespace.
    - `GafferDispatch.ExecutableNode` is no longer an alias for `GafferDispatch.TaskNode`.
    - `GafferCortex.ParameterisedHolderExecutableNode` is no longer an alias for `GafferCortex.ParameterisedHolderTaskNode`.
    - `GafferCortex` attributes can no longer be accessed via the `Gaffer` namespace.
    - `GafferCortexUI` attributes can no longer be accessed via the `GafferUI` namespace.

danieldresser-ie commented:
All sounds reasonable.  Kinda wish it was getting rolled out at IE when Andrew was around to make some of the calls on hunting down old reference boxes vs local copying of compatibility configs, but we'll figure something out.

johnhaddon commented:
> Kinda wish it was getting rolled out at IE when Andrew was around

Yeah, that's understandable. This is just for master, which I don't anticipate being released until well after Andrew's return, so it may be that decisions won't be needed until then anyway.

themissingcow commented:
> All sounds reasonable. Kinda wish it was getting rolled out at IE when Andrew was around to make some of the calls on hunting down old reference boxes vs local copying of compatibility configs, but we'll figure something out.

Is there some `grep` for the serialiser version trick that could be done?

johnhaddon commented:
Looks like I've fallen foul of my own cleanup - I'll push some fixes tomorrow.

johnhaddon commented:
OK, I've pushed with some fixups. Let's see what the CI says now.

themissingcow commented on startup/GafferUI/gafferDispatchUICompatibility.py:
```diff
@@ -1,46 +0,0 @@
-for name in dir( GafferDispatchUI ) :
-	if name.endswith( "__" ) :
-		continue
-	setattr( GafferUI, name, getattr( GafferDispatchUI, name ) )
```
https://github.com/GafferHQ/gaffer/blob/e21ae3a2c6ff98437b24056f905e6b365eaba21f/python/GafferDispatchUI/DispatcherUI.py#L133
https://github.com/GafferHQ/gaffer/blob/e21ae3a2c6ff98437b24056f905e6b365eaba21f/python/GafferDispatchUI/WedgeUI.py#L67
https://github.com/GafferHQ/gaffer/blob/e21ae3a2c6ff98437b24056f905e6b365eaba21f/python/GafferDispatchUI/WedgeUI.py#L71

johnhaddon replied:
Fixed in 276b139.

johnhaddon commented on startup/Gaffer/gafferCortexCompatibility.py:
```diff
@@ -1,52 +0,0 @@
-	for name in dir( GafferCortex ) :
-		if name.endswith( "__" ) :
-			continue
-		setattr( Gaffer, name, getattr( GafferCortex, name ) )
```
Fixed in 49629cc.

johnhaddon commented on startup/Gaffer/expressionCompatibility.py:
```diff
@@ -1,69 +0,0 @@
-def __getItem( self, name ) :
-
-	if name == "engine" and name not in self :
-		self["engine"] = Gaffer.StringPlug( flags = Gaffer.Plug.Flags.Default & ~Gaffer.Plug.Flags.Serialisable )
```
Fixed in 93588b0.

johnhaddon commented on startup/GafferDispatch/requirementsCompatibility.py:
```diff
@@ -1,88 +0,0 @@
-
-		if isinstance( key, str ) and __isPreTasksPlug( self ) :
-			if key not in self :
-				key = key.replace( "requirement", "preTask" )
```
Fixed in 5990c37.

johnhaddon commented:
Thanks for catching all those oversights Tom, I should have got them myself. I've addressed all your inline comments, and after a bit more checking also added the additional fixups in 
4d6c300 and 5073cd9.

When/if you're happy, let me know, and I'll rebase to squash and fix the conflicts.

johnhaddon commented:
Rebased - will merge once CI goes through.
